src/intervals/bed.rl

```diff
@@ -0,0 +1,278 @@
+                         thick_first=Nullable{Int}(),
+                         thick_last=Nullable{Int}(),
+                         item_rgb=Nullable{Float32}(),
+                         block_count=Nullable{Int}(),
+                         block_sizes=Nullable{Vector{Int}}(),
+                         block_firsts=Nullable{Vector{Int}}())
+        return new(name, score, thick_first, thick_last, item_rgb,
+                   block_count, block_sizes, block_firsts)
+    end
+end
+
+
+@doc """
+An `Interval` with associated metadata from a BED file.
+""" ->
+typealias BEDInterval Interval{BEDMetadata}
+
+
+module BEDParserImpl
+
+import Bio.Intervals: Strand, STRAND_NA, BEDInterval, BEDMetadata
+import Bio.Ragel
+using Docile, Switch, Compat, Color
+export BEDParser, takevalue!
+
+
+%%{
+    machine bed;
+
+    action yield {
+        yield = true
+        # // fbreak causes will cause the pushmark action for the next seqname
+        # // to be skipped, so we do it here
+        Ragel.@pushmark!
+        fbreak;
+    }
+
+    action count_line { input.state.linenum += 1 }
+    action pushmark { Ragel.@pushmark! }
+
+    action seqname     { input.seqname      = Ragel.@bytestring_from_mark! }
+    action first       { input.first        = Ragel.@int64_from_mark! }
+    action last        { input.last         = Ragel.@int64_from_mark! }
+    action name        { input.name         = Nullable{String}(Ragel.@bytestring_from_mark!) }
+    action score       { input.score        = Ragel.@int64_from_mark! }
+    action strand      { input.strand       = convert(Strand, Ragel.@char) }
+    action thick_first { input.thick_first  = Ragel.@int64_from_mark! }
+    action thick_last  { input.thick_last   = Ragel.@int64_from_mark! }
+    action item_rgb_r  { input.red = input.green = input.blue = (Ragel.@int64_from_mark!) / 255.0 }
+    action item_rgb_g  { input.green        = (Ragel.@int64_from_mark!) / 255.0 }
+    action item_rgb_b  { input.blue         = (Ragel.@int64_from_mark!) / 255.0 }
+    action item_rgb    { input.item_rgb     = RGB{Float32}(input.red, input.green, input.blue ) }
+    action block_count { input.block_count  = Ragel.@int64_from_mark! }
+    action block_size {
+        if isnull(input.block_sizes)
+            input.block_sizes = Array(Int, 0)
+        end
+        push!(get(input.block_sizes), (Ragel.@int64_from_mark!))
+    }
+    action block_first {
+        if isnull(input.block_firsts)
+            input.block_firsts = Array(Int, 0)
+        end
+        push!(get(input.block_firsts), (Ragel.@int64_from_mark!))
+    }
+
+    newline      = '\r'? '\n'     >count_line;
+    hspace       = [ \t\v];
+    blankline    = hspace* newline;
+
+    seqname      = [ -~]*   >pushmark     %seqname;
+    first        = digit+   >pushmark     %first;
+    last         = digit+   >pushmark     %last;
+    name         = [ -~]*   >pushmark     %name;
+    score        = digit+   >pushmark     %score;
+    strand       = [+\-]    >strand;
```

Comment:
This may be too strict to parse a wild BED file.
The [*rtracklayer*](http://www.bioconductor.org/packages/release/bioc/html/rtracklayer.html) package of Bioconductor can read `*` and `.` strand.